lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,472 @@
+      : valOrOp(valOrOp), fieldID(fieldID) {}
+
+  bool operator==(const CombNodeImpl &other) const {
+    return valOrOp == other.valOrOp && fieldID == other.fieldID;
+  }
+
+  CombNode *child_begin() { return children.begin(); }
+  CombNode *child_end() { return children.end(); }
+
+  void addChild(CombNode child) { children.emplace_back(child); }
+  void clearChilds() { children.clear(); }
+
+  ValueOrOpType getValOrOp() const { return valOrOp; }
+  unsigned getFieldID() const { return fieldID; }
+
+  llvm::hash_code hash_value() const {
+    return llvm::hash_combine(valOrOp.getOpaqueValue(), fieldID);
+  }
+
+private:
+  /// The underneath value or operation of this node.
+  ValueOrOpType valOrOp;
+
+  /// The field ID this node is referencing to.
+  unsigned fieldID;
+
+  /// Store all children of this node in the graph.
+  SmallVector<CombNode, 4> children;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// CombGraph class
+//===----------------------------------------------------------------------===//
+
+namespace {
+// This is a combinational graph contains a set of nodes.
+class CombGraph {
+public:
+  /// Get a node from the graph if it exists. Otherwise, allocate a new node and
+  /// add to the graph.
+  CombNode getOrAddNode(ValueOrOpType valOrOp, unsigned fieldID = 0) {
+    auto node = CombNodeImpl(valOrOp, fieldID);
+    auto it = nodes.find(&node);
+    if (it != nodes.end())
+      return *it;
+    auto newNode = new (allocator) CombNodeImpl(std::move(node));
+    nodes.insert(newNode);
+    return newNode;
+  }
+
+  /// Get a node from the graph if it exists.
+  CombNode getNode(ValueOrOpType valOrOp, unsigned fieldID = 0) {
+    auto node = CombNodeImpl(valOrOp, fieldID);
+    auto it = nodes.find(&node);
+    if (it != nodes.end())
+      return *it;
+    return nullptr;
+  }
+
+private:
+  /// For allocating memories for combinational nodes.
+  llvm::BumpPtrAllocator allocator;
+
+  /// A set of combinational nodes.
+  DenseSet<CombNode> nodes;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// GraphTraits on CombNode
+//===----------------------------------------------------------------------===//
+
+namespace llvm {
+template <>
+struct GraphTraits<CombNode> {
+  using NodeRef = CombNode;
+  using ChildIteratorType = NodeRef *;
+
+  static NodeRef getEntryNode(NodeRef node) { return node; }
+
+  static inline ChildIteratorType child_begin(NodeRef node) {
+    return node->child_begin();
+  }
+  static inline ChildIteratorType child_end(NodeRef node) {
+    return node->child_end();
+  }
+};
+} // namespace llvm
+
+//===----------------------------------------------------------------------===//
+// Detector class
+//===----------------------------------------------------------------------===//
+
+using CombPaths = SmallDenseMap<int64_t, SmallVector<int64_t>>;
```

Comment:
I'm not sure what is really being modeled here, but this is a pretty expensive data structure.  Can you add comments saying more of what the key and value represent?  Also, why a sparse map from an integer to a vector?  Can this be a dense mapping (i.e. an array)?

Also, SmallVector<T> is really designed to be used on the stack.  What is the expected length of this?  It is hard for me to tell with more context on how this is being used.

Reply:
Sorry for the confusing :sweat_smile: This is the data structure for representing combinational paths between IOs. I'd consider in most cases, this should be a very sparse map. I'm imagining a module with AXI interfaces, say 50 IOs in total, but none or only several of them are connected combinationally. Any thoughts on this? @drom @seldridge